blockchain/blockchain.go

```diff
@@ -41,22 +42,66 @@ func DecodeBlockChain(r io.Reader) *BlockChain {
 	return &bc
 }
 
-// ValidTransaction checks whether a transaction is valid, assuming the blockchain is valid.
+// ValidTransaction checks whether a transaction is valid, assuming the
+// blockchain is valid.
 func (bc *BlockChain) ValidTransaction(t *Transaction) bool {
+
 	// Find the transaction input (I) in the chain (by hash)
+	var I *Transaction
+	inputBlock := bc.Blocks[t.Input.BlockNumber]
+	for _, transaction := range inputBlock.Transactions {
+		if t.Input.Hash == HashSum(transaction) {
+			I = transaction
+		}
+	}
+	if I == nil {
+		return false
+	}
+
 	// Check that output to sender in I is equal to outputs in T
+	var inAmount uint64
+	for _, output := range I.Outputs {
+		if output.Recipient == t.Sender {
+			inAmount += output.Amount
+		}
+	}
+	var outAmount uint64
+	for _, output := range t.Outputs {
+		outAmount += output.Amount
+	}
+	if inAmount != outAmount {
+		return false
+	}
+
 	// Verify signature of T
-	return false
+	hash := HashSum(t.TxBody)
+	if !ecdsa.Verify(t.Sender.Key(), hash.Marshal(), t.Sig.R, t.Sig.S) {
+		return false
+	}
+
+	// Validate chain from input block to last block.
+	return true
 }
 
 // ValidBlock checks whether a block is valid
 func (bc *BlockChain) ValidBlock(b *Block) bool {
+	// Check that block number is one greater than last block
+	lastBlock := bc.Blocks[b.BlockNumber-1]
+	if lastBlock.BlockNumber != b.BlockNumber-1 {
+		return false
+	}
+
+	// Verify every Transaction in the block.
 	for _, t := range b.Transactions {
```

Comment:
I feel like setting the `Transaction.Input` to some special value is a good way to go. Perhaps then we could get away with treating it as a normal block during verification (ie. I'd like to avoid special cases in the `VerifyX` functions). By convention, we could put the block reward in `Transactions[0]`.